api/docs/config.md
# Configuration

## Index

- [Index](#index);
- [About](#about);
- [Schema](#schema);
- [Minimal Config](#minimal-config).

## About

This document contains detailed description of configuration schema.

## Schema

API configuration is stored in `config.json` file in `/api` folder.
By default this file doesn't exist and you have to create it on your
own. It has the following available options (use of any unsupported
option is considered an error):

| Field Path                    | Type          | Default Value | Required | Description                                                                        |
|-------------------------------|---------------|---------------|----------|------------------------------------------------------------------------------------|
| `http.prefix`                 | Path `string` | `"/api"`      | No       | HTTP path prefix of all API-requests                                               |
| `http.host`                   | `string`      | `"localhost"` | No       | HTTP server address                                                                |
| `http.port`                   | Port `number` | `8000`        | No       | HTTP server port                                                                   |
| `http.socketPath`             | Path `string` | -             | No       | HTTP server Unix-socket path. When used `api.port` and `api.host` is ignored       |
| `mysql.database`              | `string`      | `poms`        | No       | Name of database to use                                                            |
| `mysql.host`                  | `string`      | `"localhost"` | No       | MySQL server address                                                               |
| `mysql.port`                  | Port `number` | `3306`        | No       | MySQL server port                                                                  |
| `mysql.socketPath`            | Path `string` | -             | No       | MySQL server Unix-socket path. When used `mysql.host` and `mysql.port` are ignored |
| `mysql.login`                 | `string`      | -             | Yes (1)  | MySQL user login                                                                   |
| `msyql.password`              | `string`      | -             | Yes (1)  | MySQL user password                                                                |
| `mysql.init.login`            | `string`      | -             | No  (2)  | MySQL initialization user login                                                    |
| `mysql.init.password`         | `string`      | -             | No  (2)  | MySQL initialization user password                                                 |
| `mysql.serve.login`           | `string`      | -             | No  (2)  | MySQL serving user login                                                           |
| `mysql.serve.password`        | `string`      | -             | No  (2)  | MySQL serving user password                                                        |
| `logic.createAdmin`           | `boolean`     | `true`        | No       | Create default admin account on database initialization                            |
| `logic.validateTables`        | `boolean`     | `true`        | No       | Enables database tables validation                                                 |
| `logic.recreateInvalidTables` | `boolean`     | `false`       | No       | Enables dropping of invalid tables and their recreation                            |
| `logic.reconnectInterval`     | `number`      | `5`           | No       | Number of seconds between automatic reconnections to the database                  |
| `logic.maxTokens`             | `number`      | `10`          | No       | Maximum number of tokens per user                                                  |
| `logic.maxCNames`             | `number`      | `5`           | No       | Maximum number of canonical names per user                                         |

 1) Are required if any of the following options are omitted:

    - `mysql.init.login`;
    - `mysql.init.password`;
    - `mysql.serve.login`;
    - `mysql.serve.password`.

 2) Are required if `mysql.login` and `mysql.password` are omitted.

## Minimal Config

Minimal valid config is following:

```json
{
    "mysql": {
        "login":    "your MySQL user login",
        "password": "your MySQL user password"
    }
}
```


server/docs/config.md
# Configuration

## Table of Contents

- [Table of Contents](#table-of-contents);
- [About](#about);
- [Schema](#schema);
- [Placeholders](#placeholders);
- [Minimal Config](#minimal-config).

## About

This document contains detailed description of configuration schema.

## Schema

Server configuration is all stored in `poms-config.json` file. This file can be stored anywhere
up in the filesystem tree relative to server launch directory. Server will find it and set it's
directory as it's working directory. By default configuration file isn't created, you must create
it on your own. All available configuration options are listed below (use of any unsupported option
is considered an error):

| Field Path                    | Type          | Default Value                | Required | Description                                                                        |
|-------------------------------|---------------|------------------------------|----------|------------------------------------------------------------------------------------|
| `http.apiPrefix`              | Path `string` | `"/api"`                     | No       | HTTP path prefix of all API-requests                                               |
| `http.host`                   | `string`      | `"localhost"`                | No       | HTTP server address                                                                |
| `http.port`                   | Port `number` | `8000`                       | No       | HTTP server port                                                                   |
| `http.socketPath`             | Path `string` | -                            | No       | HTTP server Unix-socket path. When used `api.port` and `api.host` is ignored       |
| `http.serveStatic`            | `boolean`     | `true`                       | No       | If `true` HTTP server will serve static content from `http.staticPath`             |
| `http.staticPath`             | Path `string` | `<SITE_PATH>/build`          | No       | Path to static content HTTP server to serve                                        |
| `http.error404Path`           | Path `string` | `<SITE_PATH>/build/404.html` | No       | Path to 404-error page                                                             |
| `http.error500Path`           | Path `string` | `<SITE_PATH>/build/500.html` | No       | Path to 500-error page                                                             |
| `mysql.database`              | `string`      | `poms`                       | No       | Name of database to use                                                            |
| `mysql.host`                  | `string`      | `"localhost"`                | No       | MySQL server address                                                               |
| `mysql.port`                  | Port `number` | `3306`                       | No       | MySQL server port                                                                  |
| `mysql.socketPath`            | Path `string` | -                            | No       | MySQL server Unix-socket path. When used `mysql.host` and `mysql.port` are ignored |
| `mysql.login`                 | `string`      | -                            | Yes (1)  | MySQL user login                                                                   |
| `msyql.password`              | `string`      | -                            | Yes (1)  | MySQL user password                                                                |
| `mysql.init.login`            | `string`      | -                            | No  (2)  | MySQL initialization user login                                                    |
| `mysql.init.password`         | `string`      | -                            | No  (2)  | MySQL initialization user password                                                 |
| `mysql.serve.login`           | `string`      | -                            | No  (2)  | MySQL serving user login                                                           |
| `mysql.serve.password`        | `string`      | -                            | No  (2)  | MySQL serving user password                                                        |
| `mysql.validateTables`        | `boolean`     | `true`                       | No       | Enables database tables validation                                                 |
| `mysql.recreateInvalidTables` | `boolean`     | `false`                      | No       | Enables dropping of invalid tables and their recreation                            |
| `mysql.reconnectInterval`     | `number`      | `5`                          | No       | Number of seconds between automatic reconnections to the database                  |
| `logic.createAdmin`           | `boolean`     | `true`                       | No       | Create default admin account on database initialization                            |
| `logic.maxTokens`             | `number`      | `10`                         | No       | Maximum number of tokens per user                                                  |
| `logic.maxNicknames`          | `number`      | `5`                          | No       | Maximum number of nicknames per user                                               |
| `logic.buildStatic`           | `boolean`     | `true`                       | No       | Build static conent if `http.staticPath` is empty or doesn't exits                 |
| `logic.buildStaticPath`       | Path `string` | `<SITE_PATH>`                | No       | Path for running `npm run build` to build static content if needed                 |
| `logic.openBrowser`           | `boolean`     | `true`                       | No       | If `true` opens browser after server start                                         |

 1) Are required if any of the following options are omitted:

    - `mysql.init.login`;
    - `mysql.init.password`;
    - `mysql.serve.login`;
    - `mysql.serve.password`.

 2) Are required if `mysql.login` and `mysql.password` are omitted.

## Placeholders

Configuration options representing filesystem paths can contain placeholders
starting with `<` and ending with `>`. You can escape `<` using `\` like this: `\<`.

Those are all available placeholders:

- `<POM_PATH_PATH>` - path to the POMS root directory;
- `<PLUGIN_PATH>`   - path to the Bukkit plugin directory;
- `<SERVER_PATH>`   - path to the server directory;
- `<SITE_PATH>`     - path to the site directory.

## Minimal Config

Minimal valid config is following:

```json
{
    "mysql": {
        "login":    "your MySQL user login",
        "password": "your MySQL user password"
    }
}
```


server/src/config.ts
import { promises as fsp          } from "fs"
import { normalize, join, dirname } from "path"
import { Logger                   } from "winston"
import { DeepReadonly             } from "util/type"
import { deepAssign               } from "./util/object"

import LoggedError from "./util/LoggedError"

import * as o from "./util/object"

export interface ConfigJSON {
    http?: {
        apiPrefix?:    string
        host?:         string
        port?:         number
        socketPath?:   string
        serveStatic?:  boolean
        staticPath?:   string
        error404Path?: string
        error500Path?: string
    }

    mysql: {
        database?:              string
        host?:                  string
        port?:                  number
        socketPath?:            string
        login?:                 string
        password?:              string
        validateTables?:        boolean
        recreateInvalidTables?: boolean
        reconnectInterval?:     number

        init?: {
            login?:    string
            password?: string
        }

        serve?: {
            login?:    string
            password?: string
        }
    }

    logic?: {
        createAdmin?:           boolean
        maxTokens?:             number
        maxNicknames?:          number
        buildStatic?:           boolean
        buildStaticPath?:       string
        openBrowser?:           boolean
    }
}

export interface ReadConfigFromFileOptions {
    path?:   string
    logger?: Logger
}

export default class Config {
    static readonly POMS_PATH                             = dirname(dirname(__dirname)) // this file is in /server/src/
    static readonly PLUGIN_PATH                           = join(this.POMS_PATH, "plugin")
    static readonly SERVER_PATH                           = join(this.POMS_PATH, "server")
    static readonly SITE_PATH                             = join(this.POMS_PATH, "site")

    static readonly FILE_NAME                             = "poms-config.json"

    static readonly DEFAULT_PATH                          = this.FILE_NAME

    static readonly DEFAULT_HTTP_API_PREFIX               = "/api"
    static readonly DEFAULT_HTTP_HOST                     = "localhost"
    static readonly DEFAULT_HTTP_PORT                     = 8000
    static readonly DEFAULT_HTTP_SERVE_STATIC             = true
    static readonly DEFAULT_HTTP_STATIC_PATH              = this.placehold("<SITE_PATH>/build")
    static readonly DEFAULT_HTTP_ERROR_404_PATH           = this.placehold("<SITE_PATH>/build/404.html")
    static readonly DEFAULT_HTTP_ERROR_500_PATH           = this.placehold("<SITE_PATH>/build/500.html")

    static readonly DEFAULT_MYSQL_DATABASE                = "poms"
    static readonly DEFAULT_MYSQL_HOST                    = "localhost"
    static readonly DEFAULT_MYSQL_PORT                    = 3306
    static readonly DEFAULT_MYSQL_VALIDATE_TABLES         = true 
    static readonly DEFAULT_MYSQL_RECREATE_INVALID_TABLES = false
    static readonly DEFAULT_MYSQL_RECONNECT_INTERVAL      = 5

    static readonly DEFAULT_LOGIC_CREATE_ADMIN            = true
    static readonly DEFAULT_LOGIC_MAX_TOKENS              = 10
    static readonly DEFAULT_LOGIC_MAX_NICKNAMES           = 5
    static readonly DEFAULT_LOGIC_BUILD_STAITC            = true
    static readonly DEFAULT_LOGIC_BUILD_STAITC_PATH       = this.placehold("<SITE_PATH>")
    static readonly DEFAULT_LOGIC_OPEN_BROWSER            = true

    readonly read: DeepReadonly<ConfigJSON>
    readonly path: string

    static placehold(path: string): string {
        const splits = split(path)

        return splits.map(s => {
            if (!s.isPlaceholder)
                return s.text

            switch (s.text) {
                case "POMS_PATH":
                    return this.POMS_PATH

                case "PLUGIN_PATH":
                    return this.PLUGIN_PATH

                case "SERVER_PATH":
                    return this.SERVER_PATH

                case "SITE_PATH":
                    return this.SITE_PATH

                default:
                    return `<${s.text}>`
            }
        }).join("")

        type Split = {
            text:           string
            isPlaceholder?: boolean
        }

        function split(path: string): Split[] {
            const rawSplits = path.split(/(?<!\\)</)
            const splits    = [] as Split[]

            splits.push({ text: rawSplits[0] })

            for (let i = 1; i < rawSplits.length; ) {
                const rawSplit = rawSplits[i]

                if (rawSplit.includes(">")) {
                    const endIndex = rawSplit.indexOf(">")

                    splits.push({
                        text:          rawSplit.slice(0, endIndex),
                        isPlaceholder: true
                    })

                    splits.push({ text: rawSplit.slice(endIndex + 1) })

                    i += 2

                    continue
                }

                splits.push({ text: rawSplit })

                ++i
            }

            return splits
        }
    }

    static async readFromFile(path?: string, logger?: Logger): Promise<Config> {
        path = path ?? await this.findConfig(logger)

        logger?.info("Reading config...")

        const json = await this.readJSON(path, logger)

        this.validateJSON(json, logger)

        const config = new Config(json, path)

        logger?.info("Done")

        return config
    }

    static async findConfig(logger?: Logger): Promise<string> {
        logger?.info("Searching for configuration file...")

        let dir = process.cwd()

        while (true) {
            try {
                const files = await fsp.readdir(dir)

                if (files.includes(this.FILE_NAME)) {
                    const path = join(dir, this.FILE_NAME)
                    logger?.info(`Configuration file found at ${path}`)
                    return path
                }
            } catch {}

            const newDir = dirname(dir)

            if (dir === newDir)
                throw new Error("Configuration file not found")

            dir = newDir
        }
    }

    private static async readJSON(path: string, logger?: Logger): Promise<any> {
        try {
            const buffer = await fsp.readFile(path)
            const string = buffer.toString();

            return JSON.parse(string)
        } catch (error) {
            let message: string = `Failed to open configuration file ${path}`

            if (error instanceof SyntaxError)
                message = `Configuration file ${path} is a malformed json`
            else if (error instanceof Error)
                switch ((error as any).code) {
                    case "ENOENT":
                        message = `Configuration file ${path} not found`, { cause: error }
                        break

                    case "EISDIR":
                        message = `Configuration file ${path} is a directory`
                        break

                    case "EPERM":
                        message = `You have no permission to open configuration file ${path}`
                }

            logger?.error(message)

            throw LoggedError.forward(error, logger)
        }
    }

    private static validateJSON(json: any, logger?: Logger) {
        const result = o.validate(json, {
            fields: [
                // API
                { path: "http.apiPrefix",              type: "string"  },
                { path: "http.host",                   type: "string"  },
                { path: "http.port",                   type: "number"  },
                { path: "http.socketPath",             type: "string"  },
                { path: "http.serveStatic",            type: "boolean" },
                { path: "http.staticPath",             type: "string"  },
                { path: "http.error404Path",           type: "string"  },
                { path: "http.error500Path",           type: "string"  },

                // MySQL
                { path: "mysql.database",              type: "string"  },
                { path: "mysql.host",                  type: "string"  },
                { path: "mysql.port",                  type: "number"  },
                { path: "mysql.socketPath",            type: "string"  },
                { path: "mysql.login",                 type: "string"  },
                { path: "mysql.password",              type: "string"  },
                { path: "mysql.validateTables",        type: "boolean" },
                { path: "mysql.recreateInvalidTables", type: "boolean" },
                { path: "mysql.reconnectInterval",     type: "number"  },
                { path: "mysql.init.login",            type: "string"  },
                { path: "mysql.init.password",         type: "string"  },
                { path: "mysql.serve.login",           type: "string"  },
                { path: "mysql.serve.password",        type: "string"  },

                // Logic
                { path: "logic.createAdmin",           type: "boolean" },
                { path: "maxTokens",                   type: "number"  },
                { path: "maxNicknames",                type: "number"  },
                { path: "buildStatic",                 type: "boolean" },
                { path: "buildStaticPath",             type: "string"  },
                { path: "openBrowser",                 type: "boolean" },
            ]
        })

        switch (result.error) {
            case "EXCESS":
                throw LoggedError.fromMessage(`Found unknown configuration option "${result.path}"`, logger)

            case "MISSING":
                throw LoggedError.fromMessage(`Missing required configuration option "${result.path}"`, logger)

            case "TYPE_MISMATCH":
                throw LoggedError.fromMessage(`Configuration option "${result.path}" must be of ${result.expected} type but it's of ${result.got} type`, logger)
        }

        this.validateJSONMysqlCredentials(json)
        this.validateJSONPortFields(json)
    }

    private static validateJSONMysqlCredentials(json: any, logger?: Logger) {
        if ((json.mysql?.login        == null || json.mysql?.password        == null)
         && (json.mysql?.init?.login  == null || json.mysql?.init?.password  == null
          || json.mysql?.serve?.login == null || json.mysql?.serve?.password == null)) {
            let message: string

            if (json.mysql?.login == json.mysql?.password)
                message = 'Missing "mysql.login" and "mysql.password" configuration options'
            else if (json.mysql?.login == null)
                message = 'Missing "mysql.login" configuration option'
            else
                message = 'Missing "mysql.password" configuration option'

            throw LoggedError.fromMessage(message, logger)
        }
    }

    private static validateJSONPortFields(json: any, logger?: Logger) {
        this.validateJSONPortField(json, "api.port",   logger)
        this.validateJSONPortField(json, "mysql.port", logger)
    }

    private static validateJSONPortField(json: any, path: string, logger?: Logger) {
        const port = o.getField(json, path)

        if (port == null)
            return

        if (!Number.isInteger(port) || port < 0 || port > 65535)
            throw LoggedError.fromMessage(`Configuration option "${path}" must be a valid port number (an unsigned integer in range [0, 65535])`, logger)
    }

    constructor(json: ConfigJSON, path?: string) {
        const read = deepAssign({}, json)

        if (read.http != null) {
            read.http.apiPrefix       = normalize(`/${read.http.apiPrefix ?? ""}`)
            read.http.socketPath   = preparePath(read.http.socketPath)
            read.http.staticPath   = preparePath(read.http.staticPath)
            read.http.error404Path = preparePath(read.http.error404Path)
        }

        read.mysql.socketPath = preparePath(read.mysql.socketPath)

        if (read.logic != null)
            read.logic.buildStaticPath = preparePath(read.logic.buildStaticPath)

        this.read = read
        this.path = path ?? Config.DEFAULT_PATH

        function preparePath(path: string | undefined): typeof path {
            if (path == null)
                return undefined

            return Config.placehold(normalize(path))
        }
    }

    get mysqlHost() {
        return this.read.mysql?.host ?? Config.DEFAULT_MYSQL_HOST
    }

    get mysqlPort() {
        return this.read.mysql?.port ?? Config.DEFAULT_MYSQL_PORT
    }

    get mysqlDatabase(): string {
        return this.read.mysql?.database ?? Config.DEFAULT_MYSQL_DATABASE
    }

    get mysqlAddress(): string {
        const mysql = this.read.mysql

        if (mysql?.socketPath != null)
            return `unix:${mysql!.socketPath}`

        const host = mysql?.host ?? Config.DEFAULT_MYSQL_HOST
        const port = mysql?.port ?? Config.DEFAULT_MYSQL_PORT

        return `${host}:${port}`
    }

    get httpApiAddress(): string {
        const api    = this.read.http
        const host   = api?.host   ?? Config.DEFAULT_HTTP_HOST
        const prefix = api?.apiPrefix ?? Config.DEFAULT_HTTP_API_PREFIX

        if (api?.socketPath != null)
            return `http://unix:${api.socketPath}${prefix}/`

        const port = api?.port ?? Config.DEFAULT_HTTP_PORT

        return `http://${host}:${port}${prefix}/`
    }

    get httpAddress(): string {
        const api    = this.read.http
        const host   = api?.host ?? Config.DEFAULT_HTTP_HOST

        if (api?.socketPath != null)
            return `http://unix:${api.socketPath}/`

        const port = api?.port ?? Config.DEFAULT_HTTP_PORT

        return `http://${host}:${port}/`
    }

    get mysqlUseInitUser(): boolean {
        const mysql = this.read.mysql

        return mysql?.init?.login    != null 
            && mysql?.init?.password != null
    }

    get mysqlUseServeUser(): boolean {
        const mysql = this.read.mysql

        return mysql?.serve?.login    != null 
            && mysql?.serve?.password != null
    }

    get httpPrefix(): string {
        return this.read.http?.apiPrefix ?? Config.DEFAULT_HTTP_API_PREFIX
    }

    get httpHost(): string {
        return this.read.http?.host ?? Config.DEFAULT_HTTP_HOST
    }

    get httpPort(): number {
        return this.read.http?.port ?? Config.DEFAULT_HTTP_PORT
    }

    get httpServeStatic(): boolean {
        return this.read.http?.serveStatic ?? Config.DEFAULT_HTTP_SERVE_STATIC
    }

    get httpStaticPath(): string {   
        return this.read.http?.staticPath ?? Config.DEFAULT_HTTP_STATIC_PATH
    }

    get logicCreateAdmin(): boolean {
        return this.read.logic?.createAdmin ?? Config.DEFAULT_LOGIC_CREATE_ADMIN
    }

    get mysqlValidateTables(): boolean {
        return this.read.mysql?.validateTables ?? Config.DEFAULT_MYSQL_VALIDATE_TABLES
    }

    get mysqlRecreateInvalidTables(): boolean {
        return this.read.mysql?.recreateInvalidTables ?? Config.DEFAULT_MYSQL_RECREATE_INVALID_TABLES
    }

    get mysqlReconnectInterval(): number {
        return this.read.mysql?.reconnectInterval ?? Config.DEFAULT_MYSQL_RECONNECT_INTERVAL
    }

    get logicMaxTokens(): number {
        return this.read.logic?.maxTokens ?? Config.DEFAULT_LOGIC_MAX_TOKENS
    }

    get logicMaxCNames(): number {
        return this.read.logic?.maxTokens ?? Config.DEFAULT_LOGIC_MAX_NICKNAMES
    }

    get logicBuildStatic(): boolean {
        return this.read.logic?.buildStatic ?? Config.DEFAULT_LOGIC_BUILD_STAITC
    }

    get logicBuildStaticPath(): string {
        return this.read.logic?.buildStaticPath ?? Config.DEFAULT_LOGIC_BUILD_STAITC_PATH
    }

    get httpError404Path(): string {
        return this.read.http?.error404Path ?? Config.DEFAULT_HTTP_ERROR_404_PATH
    }

    get httpError500Path(): string {
        return this.read.http?.error500Path ?? Config.DEFAULT_HTTP_ERROR_500_PATH
    }

    get logicOpenBrowser(): boolean {
        return this.read.logic?.openBrowser ?? Config.DEFAULT_LOGIC_OPEN_BROWSER
    }
}